gpu-aware-scheduling/pkg/gpuscheduler/node_resource_cache.go

```diff
@@ -0,0 +1,551 @@
+	"k8s.io/klog/v2"
+)
+
+const (
+	add              = true
+	remove           = false
+	workerWaitTime   = time.Millisecond * 100
+	informerInterval = time.Second * 30
+)
+
+//nolint: gochecknoglobals // only mocked APIs are allowed as globals
+var (
+	internalCacheAPI InternalCacheAPI
+)
+
+// Errors.
+var (
+	errUnknownAction = errors.New("unknown action")
+	errHandling      = errors.New("error handling pod")
+	errBadArgs       = errors.New("bad args")
+)
+
+//nolint: gochecknoinits // only mocked APIs are allowed in here
+func init() {
+	internalCacheAPI = &realInternalCacheAPI{}
+}
+
+// Cache : basically all things cached, including the resulting resource usage statuses per card
+// Nodes cache is needed for incoming scheduling request so that not all node objects need to be
+// sent for every scheduled pod.
+// Pods cache is needed during the scheduling request so that not all pods need to be read from
+// all nodes for every scheduled pod.
+// The cache could be accessed from multiple goroutines and therefore needs concurrency protection,
+// which is achieved with a mutex.
+type Cache struct {
+	clientset             kubernetes.Interface
+	sharedInformerFactory informers.SharedInformerFactory
+	nodeLister            corev1.NodeLister
+	workQueue             workqueue.RateLimitingInterface
+	podLister             corev1.PodLister
+	annotatedPods         map[string]string
+	rwmutex               sync.RWMutex
+	nodeStatuses          map[string]nodeResources
+}
+
+// Node resources = a map of resourceMaps accessed by node card names.
+type nodeResources map[string]resourceMap
+
+const /*action*/ (
+	podUpdated = iota
+	podAdded
+	podDeleted
+	podCompleted
+)
+
+type workQueueItem struct {
+	name       string
+	ns         string
+	annotation string
+	action     int
+	pod        *v1.Pod
+}
+
+// NewCache returns a new Cache object.
+func NewCache(client kubernetes.Interface) *Cache {
+	if client == nil {
+		klog.Error("Can't create cache with nil clientset")
+
+		return nil
+	}
+
+	sharedInformerFactory := informers.NewSharedInformerFactory(client, informerInterval)
+	nodeInformer := sharedInformerFactory.Core().V1().Nodes()
+	nodeLister := nodeInformer.Lister()
+	podInformer := sharedInformerFactory.Core().V1().Pods()
+	podLister := podInformer.Lister()
+	stopChannel := signalHandler()
+
+	klog.V(L1).Info("starting shared informer factory (cache)")
+
+	go sharedInformerFactory.Start(stopChannel)
+
+	syncOk := internalCacheAPI.WaitForCacheSync(stopChannel, nodeInformer.Informer().HasSynced)
+	if syncOk {
+		klog.V(L2).Info("node cache created and synced successfully")
+	} else {
+		klog.Error("Couldn't sync clientgo cache for nodes")
+
+		return nil
+	}
+
+	syncOk = internalCacheAPI.WaitForCacheSync(stopChannel, podInformer.Informer().HasSynced)
+	if syncOk {
+		klog.V(L2).Info("POD cache created and synced successfully")
+	} else {
+		klog.Error("Couldn't sync clientgo cache for PODs")
+
+		return nil
+	}
+
+	c := Cache{
```

Comment:
so the controllers are part of the cache - but they act on the cache itself through their handler methods in order to update pod info?

Reply:
Yeah this was a kiss-approach, it is a bit of a blob. These could be separated to make roles more explicit, but the structure doesn't really become simpler at the end with that.